src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -541,101 +532,98 @@ private async ValueTask<TargetingContext> ResolveTargetingContextAsync(Cancellat
+                        }
                     }
                 }
-            }
 
-            if (featureDefinition.Allocation.Percentile != null)
-            {
-                foreach (PercentileAllocation percentile in featureDefinition.Allocation.Percentile)
+                if (evaluationEvent.FeatureDefinition.Allocation.Percentile != null)
                 {
-                    if (TargetingEvaluator.IsTargeted(
-                        targetingContext,
-                        percentile.From,
-                        percentile.To,
-                        _assignerOptions.IgnoreCase,
-                        featureDefinition.Allocation.Seed ?? $"allocation\n{featureDefinition.Name}"))
+                    foreach (PercentileAllocation percentile in evaluationEvent.FeatureDefinition.Allocation.Percentile)
                     {
-                        if (string.IsNullOrEmpty(percentile.Variant))
+                        if (TargetingEvaluator.IsTargeted(
+                            targetingContext,
+                            percentile.From,
+                            percentile.To,
+                            _assignerOptions.IgnoreCase,
+                            evaluationEvent.FeatureDefinition.Allocation.Seed ?? $"allocation\n{evaluationEvent.FeatureDefinition.Name}"))
                         {
-                            Logger?.LogWarning($"Missing variant name for percentile allocation in feature {featureDefinition.Name}");
+                            if (string.IsNullOrEmpty(percentile.Variant))
+                            {
+                                Logger?.LogWarning($"Missing variant name for percentile allocation in feature {evaluationEvent.FeatureDefinition.Name}");
 
-                            return new ValueTask<VariantDefinition>((VariantDefinition)null);
-                        }
+                                return new ValueTask<VariantDefinition>((VariantDefinition)null);
+                            }
 
-                        Debug.Assert(featureDefinition.Variants != null);
+                            Debug.Assert(evaluationEvent.FeatureDefinition.Variants != null);
 
-                        return new ValueTask<VariantDefinition>(
-                            featureDefinition
-                                .Variants
-                                .FirstOrDefault((variant) => variant.Name == percentile.Variant));
+                            evaluationEvent.AssignmentReason = AssignmentReason.Percentile;
+
+                            return new ValueTask<VariantDefinition>(
+                                evaluationEvent.FeatureDefinition
+                                    .Variants
+                                    .FirstOrDefault((variant) => variant.Name == percentile.Variant));
+                        }
                     }
                 }
             }
 
+            if (variant == null)
```

Comment:
I would suggest to let the caller decide to set the assignment reason to none.

Reply:
Actually, I don't see us ever setting `variant` to anything other than null when we first declare it. Am I missing something?